### PR TITLE
LXC Support

### PR DESCRIPTION
**Pull request by diptanu**
This PR implements support for the LXC driver. The driver is enabled only in Linux.

### CODE REVIEW

File: client/driver/lxc.go [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+	}
+
+	if err := c.Create(options); err != nil {
+		return nil, fmt.Errorf("unable to create container: %v", err)
+	}
+	if err := c.Start(); err != nil {
+		return nil, fmt.Errorf("unable to start container: %v", err)
+	}
+	// Set the resource limits
+	if err := c.SetMemoryLimit(lxc.ByteSize(task.Resources.MemoryMB) * lxc.MB); err != nil {
```
Comment (schmichael) [inline client/driver/lxc.go]: Should this come before `c.Start()`?

File: client/driver/lxc.go [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+		waitCh:         make(chan *dstructs.WaitResult, 1),
+		doneCh:         make(chan bool, 1),
+	}
+
+	go handle.run()
+
+	return &handle, nil
+}
+
+// Open opena the driver to monitor an existing LXC container
```
Comment (schmichael) [inline client/driver/lxc.go]: `// Open opena` -> `// Open the driver`

File: client/driver/lxc.go [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+	for {
+		select {
+		case <-timer.C:
+			process, err := os.FindProcess(h.initPid)
+			if err != nil {
+				h.waitCh <- &dstructs.WaitResult{Err: err}
+				return
+			}
+			if err := process.Signal(syscall.Signal(0)); err != nil {
+				h.waitCh <- &dstructs.WaitResult{}
```
Comment (schmichael) [inline client/driver/lxc.go]: Do you want to include `err` in `WaitResult`?

File: client/driver/lxc.go [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+		case <-h.doneCh:
+			h.waitCh <- &dstructs.WaitResult{}
+			return
+		}
+	}
+}
+
+func keysToVal(line string) (string, uint64, error) {
+	tokens := strings.Split(line, " ")
+	key := tokens[0]
```
Comment (schmichael) [inline client/driver/lxc.go]: Should there be a length check here to prevent a panic or if `line` doesn't contain a space is everything broken anyway?

**Comment by schmichael** [comment]
Neither of my comments are blockers. LGTM

File: client/driver/lxc.go (hunk 3 above)
Comment (diptanu) [inline client/driver/lxc.go]: @schmichael This is a cheap trick to know if a process is alive on Linux. So this is not really an error in the container, we are just checking if the init-pid of the container is still alive.

File: client/driver/lxc.go (hunk 4 above)
Comment (diptanu) [inline client/driver/lxc.go]: Yeah agreed. Adding a check here doesn't hurt.

**Comment by diptanu** [comment]
@schmichael Addressed all your comments.

**Comment by diptanu** [comment]
@dadgar I would recommend merging this PR since the basic features have been implemented. We can add support for downloading support of rootfs and MacVLAN in subsequent PRs.

File: client/driver/lxc.go [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+	if !ok {
+		return nil, fmt.Errorf("failed to find task local directory: %v", task.Name)
+	}
+	taskDirMount := fmt.Sprintf("%s local none rw,bind,create=dir", taskLocalDir)
+	allocDirMount := fmt.Sprintf("%s alloc none rw,bind,create=dir", ctx.AllocDir.SharedDir)
+
+	if err := c.SetConfigItem("lxc.mount.entry", allocDirMount); err != nil {
+		return nil, fmt.Errorf("error setting alloc dir bind mounts configuration: %v", err)
+	}
+	if err := c.SetConfigItem("lxc.mount.entry", taskDirMount); err != nil {
```
Comment (schmichael) [inline client/driver/lxc.go]: Calling `SetConfigItem("lxc.mount.entry",...` twice will append the second entry, not overwrite the first?

I couldn't tell from looking at the lxc docs.

File: client/driver/lxc.go [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+
+func (h *lxcDriverHandle) Update(task *structs.Task) error {
+	h.killTimeout = GetKillTimeout(task.KillTimeout, h.killTimeout)
+	return nil
+}
+
+func (h *lxcDriverHandle) Kill() error {
+	h.logger.Printf("[INFO] driver.lxc: shutting down container %q", h.container.Name())
+	if err := h.container.Shutdown(h.killTimeout); err != nil {
+		h.logger.Printf("[INFO] driver.lxc: shutting down container %q failed: %v", h.container.Name(), err)
```
Comment (schmichael) [inline client/driver/lxc.go]: `INFO` -> `WARN` perhaps?

File: client/driver/lxc.go (hunk 5 above)
Comment (diptanu) [inline client/driver/lxc.go]: No it won't, if you look at lxc configuration there can be multiple `lxc.mount.entry` - as many fstab entries you want.

File: client/driver/lxc.go [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/mitchellh/mapstructure"
+
+	dstructs "github.com/hashicorp/nomad/client/driver/structs"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	lxc "gopkg.in/lxc/go-lxc.v2"
+)
+
+const (
+
```
Comment (dadgar) [inline client/driver/lxc.go]: Extra line?

File: client/driver/lxc.go [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+	containerMonitorIntv = 2 * time.Second
+)
+
+var (
+	LXCMeasuredCpuStats = []string{"System Mode", "User Mode", "Percent"}
+	LXCMeasuredMemStats = []string{"RSS", "Cache", "Swap", "Max Usage", "Kernel Usage", "Kernel Max Usage"}
+)
+
+// Add the lxc driver to the list of builtin drivers
+func init() {
```
Comment (dadgar) [inline client/driver/lxc.go]: Why is this here, do it like the other drivers?

File: client/driver/lxc.go [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+	return true, nil
+}
+
+// Start starts the LXC Driver
+func (d *LxcDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle, error) {
+	var driverConfig LxcDriverConfig
+	if err := mapstructure.WeakDecode(task.Config, &driverConfig); err != nil {
+		return nil, err
+	}
+	lxcPath := lxc.DefaultConfigPath()
```
Comment (dadgar) [inline client/driver/lxc.go]: Can we set the path to be the task_dir?

File: client/driver/lxc.go (hunk 8 above)
Comment (diptanu) [inline client/driver/lxc.go]: @dadgar So that we don't have to build the driver on other platforms. This driver needs the lxc-dev package which is unavailable on other platforms.

Comment (dadgar) [inline client/driver/lxc.go]: Ah missed the build tag

File: client/driver/lxc.go (hunk 9 above)
Comment (diptanu) [inline client/driver/lxc.go]: I am not very sure if we can have multiple Config Paths on a machine, usually all the containers live in `/var/lib/lxc`. Having it in the task dir is desirable since then we can make use of disk space accounting.

Comment (ranjib) [inline client/driver/lxc.go]: LXC allows unprovileged containers, i.e. I can run nomad as nomad user & without any root privileges to spin LXC containers owned by nomad user. In such setup, generally the containers will stay at /home/nomad/.local/share/lxc (or $HOME/.local/share/lxc).. but this is configurable. I am not aware of any case where the nomad client would be running as different user /or same user with different lxc path.

Comment (diptanu) [inline client/driver/lxc.go]: So the lxc path can be configurable by the operator. I think @dadgar is suggesting using a different path per container.

File: vendor/gopkg.in/lxc/go-lxc.v2/container.go [hunk 10]
```diff
@@ -0,0 +1,1599 @@
...
+// Daemonize returns true if the container wished to be daemonized.
+func (c *Container) Daemonize() bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return bool(c.container.daemonize)
+}
+
+// WantDaemonize determines if the container wants to run daemonized.
+func (c *Container) WantDaemonize(state bool) error {
```
Comment (ranjib) [inline vendor/gopkg.in/lxc/go-lxc.v2/container.go]: I think we should enable  this + close file descriptor by default.

**Comment by ranjib** [comment]
@diptanu i took this for a spin during the weekend. Its working on my cluster. Could not get to went through all the code. I think you should enable wantdaemonize + close file descriptor by default.

File: client/driver/lxc.go (hunk 9 above)
Comment (dadgar) [inline client/driver/lxc.go]: Yes I was suggesting specific path per container. I was looking at lxc docs and it seems like you can

File: client/driver/lxc.go [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+	// Set the network type to none
+	if err := c.SetConfigItem("lxc.network.type", "none"); err != nil {
+		return nil, fmt.Errorf("error setting network type configuration: %v", err)
+	}
+
+	// Bind mount the shared alloc dir and task local dir in the container
+	taskLocalDir, ok := ctx.AllocDir.TaskDirs[task.Name]
+	if !ok {
+		return nil, fmt.Errorf("failed to find task local directory: %v", task.Name)
+	}
```
Comment (dadgar) [inline client/driver/lxc.go]: This is wrong, see #1830

File: client/driver/lxc.go [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+		}
+	}
+
+	// Start the container
+	if err := c.Start(); err != nil {
+		return nil, fmt.Errorf("unable to start container: %v", err)
+	}
+
+	// Set the resource limits
+	if err := c.SetMemoryLimit(lxc.ByteSize(task.Resources.MemoryMB) * lxc.MB); err != nil {
```
Comment (dadgar) [inline client/driver/lxc.go]: Set resources before starting?

File: client/driver/lxc.go [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+	"github.com/mitchellh/mapstructure"
+
+	dstructs "github.com/hashicorp/nomad/client/driver/structs"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	lxc "gopkg.in/lxc/go-lxc.v2"
+)
+
+const (
+	// The option that enables this driver in the Config.Options map.
+	lxcConfigOption = "driver.lxc.enable"
```
Comment (schmichael) [inline client/driver/lxc.go]: Fixed the comment.

@diptanu did you get a chance to write any docs? If not I don't mind doing it.
